terminal2048/lib/terminal2048/move.ex

```diff
@@ -0,0 +1,160 @@
+defmodule Terminal2048.Move do
+
+  def handle_move(game, direction) do
+    {game.board, game.score}
+    |> can_move(direction)
+    |> return_game(game, direction)
+  end
+
+  defp can_move({board, score}, :left),  do: {board, score} != move_left({board, score})
+  defp can_move({board, score}, :right), do: {board, score} != move_right({board, score})
+  defp can_move({board, score}, :up),    do: {board, score} != move_up({board, score})
+  defp can_move({board, score}, :down),  do: {board, score} != move_down({board, score})
+
+  defp return_game(false, game, _direction) do
+    %Terminal2048.State{ game |
+      game_state: :no_move
+    }
+  end
+
+  defp return_game(true, game, direction) do
+    {new_board, new_score} = move(game, direction)
+
+    new_board = new_board
+      |> Terminal2048.Board.place_number
+
+    %Terminal2048.State{ game |
+      game_state: update_game_state({new_board, new_score}),
+      board: new_board,
+      score: new_score
+    }
+  end
+
+  defp update_game_state({board, score}) do
+    lost_checker = [:left, :right, :up, :down]
+      |> Enum.all?(fn direction -> !can_move({board, score}, direction) end)
+
+    won_checker = board
+      |> Enum.any?(fn x -> x == 2048 end)
+
+    handle_game_state(lost_checker, won_checker)
+  end
+
+  defp handle_game_state(true, _),      do: :lost
+  defp handle_game_state(_, true),      do: :won
+  defp handle_game_state(false, false), do: :continue
+
+  defp move(game, :left),  do: move_left({game.board, game.score})
+  defp move(game, :right), do: move_right({game.board, game.score})
+  defp move(game, :up),    do: move_up({game.board, game.score})
+  defp move(game, :down),  do: move_down({game.board, game.score})
+
```

Comment:
This is clever, but I think you could have refactored this to remove the duplicate code. For example, the first two could be:

~~~ elixir
  defp move_left_right({board, score}, maybe_reverse) do
    row_with_score = board
      |> Enum.chunk_every(4)
      |> Enum.map(maybe_reverse)
      |> Enum.map(&leftward_operation(&1))

    new_board = row_with_score
      |> Enum.map(&List.first/1)
      |> Enum.map(maybe_reverse)
      |> List.flatten

    new_score = row_with_score
      |> Enum.map(&List.last/1)
      |> List.flatten
      |> Enum.sum

    {new_board, score + new_score}
  end


  defp move_left({board, score}) do
    move_left_right({board, score}, &no_reverse/1)
  end

  defp move_right({board, score}) do
   move_left_right({board, score}, &Enum.reverse/1)
  end
  
  defp no_reverse(x), do: x
~~~